proposals/0000-extension-lifecycle-framework.md

```diff
@@ -0,0 +1,108 @@
+
+## 2. Proposed Change Specification
+
+### 2.1. Categories and Lifecycle
+Language extensions will be classified into the following categories
+ * `Experimental` extensions are undergoing active development. The syntax and semantics that are enabled by the extension are likely to change regularly. It is expected that most new language extensions will begin as experimental. At the time of writing, `LinearTypes` and `OverloadedRecordUpdate` seem to be in this category.
+ 
+ * `Mature` extensions are considered to be finished and are not expected to undergo regular changes. These features can be used without worry of unexpected changes, and they are not known to contain serious design or implementation deficiencies. Any breaking change to a mature extension will be announced well in advance of the change being made, with a migration path provided if possible. Some extensions in this category might be `MultiWayIf`, `MonoLocalBinds`, and `ViewPatterns`.
+
+ * `Deprecated` extensions are considered to be design or implementation dead ends, and should probably not be used in new code. Deprecating an extension _must_ come with a statement about the longevity of the extension. This need not be a commitment to remove it at a concrete time, but is a commitment to either remove it or transition it to another state at a time stated by GHC HQ when it is marked as `Deprecated`. Any extension will be deprecated prior to removal. Deprecated extensions are expected to include `IncoherentInstances`, `OverlappingInstances`, `Rank2Types`, `RecordPuns`, and `TypeInType`.
+ 
+ * `Legacy` extensions are explicitly not recommended for new code, much like `Deprecated`, however, they are expected to be supported indefinitely. This may be to maintain an older language set, maintain a bridge to a newer feature, or otherwise keep backward compatibility. Expected legacy extensions include `CUSKs`, `DeepSubsumption`,  and `NPlusKPatterns`.
+
+Above we list expected categorizations of several extensions based on statements in the user's guide. We do not attempt to prescribe any particular categorizations as part of this proposal and simply provide a small number of expectations from the perspective of someone reading the existing documentation.
+
+The expected extension lifecycle includes the following transitions:
```

Comment:
Do we want to say something like "A Mature extension should either eventually become part of a GHCXXXX language set or be deprecated"?

Reply:
I don't think so. It seems unlikely to me that `MultiWayIf` will ever be either deprecated or part of a default language set, for example. This proposal is not about changing the set of available extensions, it's about more clearly and accessibly communicating about those that exist.

That could be an interesting proposal, but it's not what we're proposing here. Should it be?

Reply:
Some think that such a policy is desirable, but not everybody. Arnaud is in the process of shedding some light into this, but I think we can discuss it independent of the present proposal of warning against experimental extensions.

Reply:
I guess the stated scope of this proposal seems somewhat broad, so I wonder if we want to talk about other aspects of the lifecycle. We can always do that later. And once we have these people will start asking questions "how long does something have to be in Deprecated before it gets removed?" etc.

Reply:
@michaelpj The language was very careful to not specify a global deprecation time. Instead we left that on purpose to be determined at the time of deprecation for each extension. The intent is to require clear communication for a given extension but also is not so strict as to negatively impact GHC development.

Reply:
I actually don't think we should have a specific policy - language extensions are frequently part of a research program, and research can by definition fail! Research usually also has uncertain timelines, and part of evaluating work on language design is seeing how it is used in practice. It would be unfortunate for a strict policy to get in the way of that.

The lifecycle system proposed here is intended primarily to present facts about the extensions as they are at the time of a given release. The reason why we would like to propose that "deprecated" extensions have _some_ deadline for their removal is simply that we today really have two classes of extensions described as deprecated: those for which there is a real intention to remove them, and those that are considered a bad idea but may stick around forever. If I have a large, long-lived Haskell codebase, I need to react differently to the former warning than to the latter, as the former indicates that my existing code has an expiration date, while the latter is probably mostly important as a way to keep junior developers from introducing things that we would rather avoid if they've been reading old resources. It seems valuable to distinguish between these two states.

In other words, we're really proposing a simple but still informative and actionable model for describing language extensions WRT their likelihood of introducing bit-rot, along with a low-impact and inexpensive way for the compiler to help people act based on this information, not to impose a policy on all future R&D.

Perhaps the proposal should be amended to make it more clear that "expected" here is not normative, but rather descriptive of what we believe is the most likely path? Maybe the text just after the transitions should be before them?

Reply:
(BTW: I just emailed Arnaud about coordinating our work)

Reply:
For posterity: Arnaud got back to me quickly and indicated that he agrees that his work is largely orthogonal.

Reply:
(As he wrote below, but these threads can be hard to follow without a bit of redundancy)

Reply:
I've now clarified that the expected transitions are non-normative.